GODRIVER-2037 Don't clear the connection pool on client-side connect timeout errors.

matthewdale opened the pull request:
Update `topology.Server#ProcessHandshakeError` to not clear the server connection pool on client-side timeout errors that occur while creating a connection in-line with an operation. This is intended as a patch to prevent clearing the connection pool on specific identified conditions that currently cause the driver to clear the server connection pool unnecessarily and is not intended to handle a comprehensive set of errors that could occur during handshake.
Note that the changes here are likely irrelevant once all connections are established in the background with the `connectTimeoutMS` deadline ([GODRIVER-2038](https://jira.mongodb.org/browse/GODRIVER-2038)).

Possible cases:
1. `server.Connection(context.Background())`
Get a connection to `server` using `connectTimeoutMS`. The `server` connection pool will be cleared on timeout during handshake. This case is used for connections created by the `MinPoolSize` maintenance goroutine and by operations that use `context.Background()` for the timeout.
2. `server.Connection(context.WithTimeout(..., 1*time.Second))`
Get a connection to `server` using a timeout specified on an operation lower than `connectTimeoutMS`. The `server` connection pool will not be cleared on timeout during handshake. This case is used for connections created by operations that use `context.WithTimeout()` for timeouts shorter than the `connectTimeoutMS`.
3. `server.Connection(context.WithTimeout(..., 10*time.Minute))`
Get a connection to `server` using a timeout specified on an operation higher than `connectTimeoutMS`. The `server` connection pool will be cleared on timeout during handshake. This case is used for connections created by operations that use `context.WithTimeout()` for timeouts longer than `connectTimeoutMS`.
4. `server.Connection(context.WithCancel(...))`
Get a connection to `server` using a cancellation specified on an operation. The `server` connection pool will not be cleared on timeout during handshake. This case is used for connections created by operations that use `context.WithCancel()` for cancellation.

[GODRIVER-2037](https://jira.mongodb.org/browse/GODRIVER-2037)

iwysiu commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,163 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+			eventsCh := make(chan *event.PoolEvent)
+			var eventsWg sync.WaitGroup
+			eventsWg.Add(1)
+			go func() {
```
Could you have this go routine just set pool cleared instead of storing the events?

matthewdale replied:
That would technically work for checking if there was/wasn't a pool clear event. However, the test also relies on a few more behaviors provided by using the `channel`:
- Expect that there is at least 1 event emitted, otherwise fail
- Expect that no events are emitted after the events channel is closed, otherwise panic

iwysiu commented on x/mongo/driver/topology/connection.go:
```diff
@@ -142,6 +142,8 @@ func (c *connection) connect(ctx context.Context) {
 	}
 	defer close(c.connectDone)
 
+	_, isOperationDeadline := ctx.Deadline()
```
This will still be true if ctx.Deadline() is later than connectTimeout, which I don't think we want

matthewdale replied:
Per discussion in the sync meeting, we no longer use the existence of a deadline to indicate an operation-scoped timeout happened. Instead, we check for any errors on the passed in `Context`.

matthewdale commented on x/mongo/driver/topology/server.go:
```diff
@@ -290,6 +291,56 @@ func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint6
+			err = opErr.Err
+			// Handle the case where an error has been replaced by "net.errCanceled", which isn't
+			// exported and can't be compared directly. In this case, just compare the error message.
+			if err.Error() == "operation was canceled" {
```
That's correct, timeout errors (including context deadline exceeded) are now handled by type-asserting the error to a `net.Error` and calling the `Timeout()` method if available. Checking if the error string is `"operation was canceled"` only handles context cancellation.

matthewdale commented on x/mongo/driver/topology/server.go:
```diff
@@ -290,6 +291,56 @@ func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint6
 
+	isTimeout := func(err error) bool {
+		// Handle any errors that self-report as a timeout.
+		// TODO: underlyingError and the Timeout() type assert condition are copied from the
```
The logic from `os.IsTimeout()` is no longer necessary, so I've removed that code and comment.

matthewdale commented on x/mongo/driver/topology/server.go:
```diff
@@ -290,6 +291,62 @@ func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint6
+			}
+		}
+
+		// Extract any wrapped errors that implement Unwrap() to get the original error.
```
After simplifying some of the code based on the suggestions from the sync meeting, the `isTimeout` function now iteratively unwraps and checks if all errors are timeout/cancellation (similar to the behavior of `errors.Is()`).

kevinAlbs commented on x/mongo/driver/topology/server.go:
```diff
@@ -290,6 +292,34 @@ func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint6
+
+	// Ignore errors that indicate a client-side timeout occurred when using an operation-scoped
+	// deadline (i.e. not using connectTimeoutMS as the connection timeout).
+	if (ctxErr == context.DeadlineExceeded || ctxErr == context.Canceled) && isTimeout(wrappedConnErr) {
```
To check my understanding, is the intent behind the `isTimeout(wrappedConnErr)` to prevent a race?

I am not sure if this is realistic, but I imagine the following scenario where this is desirable:
- The context deadline expires just after a socket write error occurs
- The `ctxErr` is `context.DeadlineExceeded` but `wrappedConnErr` is the result of the socket write error.
- Because the socket write error is not a timeout, this proceeds to clear the pool.

kevinAlbs commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,217 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+			require.NoError(t, server.Connect(nil))
+
+			// Use the context returned by the test case ctxFn to call Connection. Contexts without
+			// deadlines (e.g. context.Background())
```
Should the second sentence be removed?

kevinAlbs commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,217 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+		{
+			desc: "Child context deadline exceeded error during handshake should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 100*time.Millisecond)
```
Minor: Since this is only testing that the context within the handshake times out, should this be `context.Background()` for consistency with the "Child context" test above? 
```suggestion
				return context.Background(), nil
```

matthewdale replied:
Using a `context.WithTimeout` and a `context.Background` are actually two different test cases where we expect the same behavior (when the error was caused by a child context). I decided to use the `context.WithTimeout` to cover the case that was handled incorrectly by the previous code (i.e. a context with deadline was used to call `Connection()` but was not the source of the timeout error).

I'll add a comment to clarify the intent of the `context.WithTimeout`.

kevinAlbs commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,217 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+		{
+			desc: "Dial errors unrelated to context timeouts should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 1*time.Second)
```
```suggestion
				return context.Background(), nil
```

matthewdale replied:
See above comment about using `context.WithTimeout` vs `context.Background` here.

I'll add a comment to clarify the intent of the `context.WithTimeout`.

kevinAlbs commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,217 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+
+			// Create a TCP listener on a random port but never call Accept on it. The TCP listener will
+			// behave like a TCP service that doesn't explicitly deny connections, but doesn't respond to
+			// them, forcing client-side timeouts instead.
```
Test cases like `Parent context deadline exceeded error during handshake should not clear the pool` are overriding the Handshaker, but not the Dialer. Wouldn't they need connection to succeed to perform the handshake? I am not sure how that is possible if this test server is not accepting connections.

kevinAlbs commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,217 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.desc, func(t *testing.T) {
+			t.Parallel()
```
Oh nice! I assume the `tc := tc` above is necessary to capture the value of `tc` in each of the parallel goroutines.

matthewdale replied:
That's correct, otherwise the test `func`s will not capture the actual range variable and will just use whatever the pointer value when they're actually run (typically the last value in the slice). Not doing that can cause some nasty bugs, like only the last test case ever being run.

kevinAlbs reviewed (commented):
Looking great!

matthewdale commented on mongo/integration/sdam_error_handling_test.go:
```diff
@@ -33,8 +33,8 @@ func TestSDAMErrorHandling(t *testing.T) {
 		mtOpts := mtest.NewOptions().
-			Topologies(mtest.ReplicaSet). // Don't run on sharded clusters to avoid complexity of sharded failpoints.
-			MinServerVersion("4.0").      // 4.0+ is required to use failpoints on replica sets.
+			Topologies(mtest.ReplicaSet, mtest.Single). // Don't run on sharded clusters to avoid complexity of sharded failpoints.
```
This was mostly to facilitate local testing so I didn't need to set up a replica set. I kept it because it may make running the tests locally easier for other people, too (and they should run fine for the standalone deployment on Evergreen).

divjotarora commented on x/mongo/driver/topology/server.go:
```diff
@@ -290,6 +292,32 @@ func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint6
+			}
+			// Handle the case where an error has been replaced by "net.errCanceled", which isn't
+			// exported and can't be compared directly. In this case, just compare the error message.
+			if err.Error() == "operation was canceled" {
```
Should we also check for `err == context.Canceled`? There are cases where we will return `context.Canceled` in some wrapped form and I don't think the current implementation handles those. See [here](https://github.com/mongodb/mongo-go-driver/blob/ad923e2619e149cea42ad905137020b8d05f019f/x/mongo/driver/topology/connection.go#L350-L350) and [here](https://github.com/mongodb/mongo-go-driver/blob/ad923e2619e149cea42ad905137020b8d05f019f/x/mongo/driver/topology/connection.go#L392-L392) for examples. In the former, we detect that the context has expired before doing the socket operation. In the latter, the context is cancelled after we've already dispatched the operation. We detect this via our background cancellation listener logic and then overwrite the error to `context.Canceled` if the listener fired.

matthewdale replied:
Yes, you're correct. Will add a condition for `context.Canceled` in addition to checking for the `net`-package-specific cancellation error.

divjotarora commented on x/mongo/driver/topology/server.go:
```diff
@@ -290,6 +292,32 @@ func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint6
+			if err.Error() == "operation was canceled" {
+				return true
+			}
+			if wrapper, ok := err.(interface{ Unwrap() error }); ok {
```
Super-nit: not sure if this is still the case, but IIRC we try to avoid if/else blocks when possible. Instead of using `break`, you can have this conditional `continue` and then `return false` outside of it. The existing `return false` outside the loop can probably be removed at that point.

matthewdale replied:
The compiler actually complains about a missing return if I remove the `return false` after the loop. Will remove the if/else and replace it with a
```go
if !ok {
    break
}
```

divjotarora commented on x/mongo/driver/topology/server.go:
```diff
@@ -290,6 +292,32 @@ func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint6
+
+	// Ignore errors that indicate a client-side timeout occurred when using an operation-scoped
+	// deadline (i.e. not using connectTimeoutMS as the connection timeout).
+	if (ctxErr == context.DeadlineExceeded || ctxErr == context.Canceled) && isTimeout(wrappedConnErr) {
```
nit: you can check `ctxErr != nil` here as `context.DeadlineExceeded` and `context.Canceled` are the only errors that can be returned from contexts.

matthewdale replied:
I considered simplifying to just `ctxErr != nil` but decided to leave it in case `ctx.Err()` is ever updated to return additional errors. After reading the description of `ctx.Err()`, it seems that its behavior is tied directly to `ctx.Done()`, so any new errors would probably indicate some kind of "stop now" condition, which we want to handle in the same way.

Will update to `ctxErr != nil`.

matthewdale commented on x/mongo/driver/topology/server.go:
```diff
@@ -290,6 +292,32 @@ func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint6
 		return
 	}
 
+	isTimeout := func(err error) bool {
```
You're right, it checks for timeout or cancellation. Will rename to `isTimeoutOrCanceled`.

kevinAlbs reviewed (approved):
Fantastic work, LGTM!

divjotarora commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,238 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+
+			// Close the events channel and expect that no more events are sent on the channel. Then
+			// wait for the events channel loop to return before inspecting the events slice.
+			close(eventsCh)
```
IIUC if the `Server` attempts to publish any more events, it will panic because the channel has been closed. Would it be more robust to first call `Server#Disconnect()` and then close the channel?

divjotarora reviewed (commented):
Looks great! Just some minor comments about testing.

matthewdale commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,238 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+			desc: "Dial errors unrelated to context timeouts should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				// Return a context with a timeout that will not be reached during the test.
+				return context.WithTimeout(context.Background(), 1*time.Minute)
```
I've added an additional test case that tests operations that use `context.Background()` in addition to operations that use a context with timeout.

matthewdale commented on mongo/integration/sdam_error_handling_test.go:
```diff
@@ -48,7 +48,7 @@ func TestSDAMErrorHandling(t *testing.T) {
 		mt.RunOpts("network errors", noClientOpts, func(mt *mtest.T) {
-			mt.Run("pool cleared on network timeout", func(mt *mtest.T) {
+			mt.Run("pool not cleared on operation-scoped network timeout", func(mt *mtest.T) {
 				// Assert that the pool is cleared when a connection created by an application operation thread
```
Good catch. Will update.

matthewdale commented on mongo/integration/sdam_error_handling_test.go:
```diff
@@ -78,16 +78,60 @@ func TestSDAMErrorHandling(t *testing.T) {
+			})
+
+			mt.Run("pool cleared on non-operation-scoped network timeout", func(mt *mtest.T) {
+				// Assert that the pool is cleared when a connection created by an application operation thread
```
Will rewrite the comment to be more specific and remove the 2nd sentence.

kevinAlbs commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,221 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+
+			// Close the events channel and expect that no more events are sent on the channel. Then
+			// wait for the events channel loop to return before inspecting the events slice.
+			server.Disconnect(context.Background())
```
It looks like the error returned needs to captured to satisfy the [sa-errcheck](https://spruce.mongodb.com/task/mongo_go_driver_static_analysis_sa_errcheck_patch_4adb6d73ca0fe53c99e11765005f37033acd6471_60d56ac30ae60671e32a3517_21_06_25_05_33_57/logs?execution=0) task on the Evergreen patch for this PR.

matthewdale replied:
Good catch. Updated with explicitly ignoring the error.

kevinAlbs commented on mongo/integration/sdam_error_handling_test.go:
```diff
@@ -78,16 +79,59 @@ func TestSDAMErrorHandling(t *testing.T) {
 				assert.True(mt, isPoolCleared(), "expected pool to be cleared but was not")
 			})
+
 			mt.RunOpts("pool cleared on non-timeout network error", noClientOpts, func(mt *mtest.T) {
```
This test appeared to timeout in [this task](https://evergreen.mongodb.com/lobster/evergreen/test/mongo_go_driver_tests_42_plus_zlib_zstd_support__version~4.4_os_ssl_40~ubuntu1804_64_go_1_16_test_replicaset_auth_ssl_patch_4adb6d73ca0fe53c99e11765005f37033acd6471_60d56ac30ae60671e32a3517_21_06_25_05_33_57/0/be7221d15d9e87e5280e416c7e66b9d9#bookmarks=0%2C17748&l=1&shareLine=11762). I assume that is timing out on the channel read, indicating a pool clear does not occur.

It is not clear to me why. Given that it is not a consistent failure, one thing that does look suspicious to me is the fact that `clearPoolChan()` is called after `mt.ResetClient`. That looks like a race to me (monitoring starts and immediately clears the pool, then the channel is cleared, then the channel is read from).

matthewdale replied:
Ah, good catch on the timeout failure. The test that timed out was actually in the `"after handshake completes"` block, which I've seen fail intermittently before and have so far been ignoring. Will look further into the timeout failure and see if other intermittent test failures in `sdam_error_handling_test.go` are related.

matthewdale replied:
I've replaced all uses of the shared `poolMonitor` and associated `clearPoolChan()` and `isPoolCleared()` functions with individual instances of the new `testPoolMonitor` type. The tests seem to reliably pass now.

kevinAlbs commented on mongo/integration/sdam_error_handling_test.go:
```diff
@@ -78,16 +79,59 @@ func TestSDAMErrorHandling(t *testing.T) {
 				assert.NotNil(mt, err, "expected InsertOne error, got nil")
 				assert.True(mt, mongo.IsTimeout(err), "expected timeout error, got %v", err)
 				assert.True(mt, mongo.IsNetworkError(err), "expected network error, got %v", err)
+				assert.False(mt, isPoolCleared(), "expected pool not to be cleared but was cleared")
```
This assertion errored in [this task](https://evergreen.mongodb.com/lobster/evergreen/test/mongo_go_driver_tests_42_plus_zlib_zstd_support__version~latest_os_ssl_40~ubuntu1804_64_go_1_16_test_replicaset_auth_nossl_patch_4adb6d73ca0fe53c99e11765005f37033acd6471_60d56ac30ae60671e32a3517_21_06_25_05_33_57/0/ac6e3fcaffebeef411d9d724df4dcfe6#bookmarks=0%2C19006&l=1&shareLine=11759).

Similar to my comment below, my guess is that there is a race in the order here:

```
mt.ResetClient(clientOpts)
clearPoolChan()
```

Though that begs the question of why this was not failing before.

matthewdale replied:
Will look into any race conditions with `poolMonitor`.

divjotarora commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,238 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
 }
 
+func TestServerConnectionTimeout(t *testing.T) {
+	testCases := []struct {
```
Yes, I do think it's worth testing. You can test the case where the parent context is cancelled by taking advantage of the fact that anonymous functions can access variables outside of their scope:

```
ctx, cancel := context.WithCancel(context.Background()) // parent context passed to connect()
dialer := func(Dialer) Dialer {
    // Inside the returned DialerFunc, call cancel()
}
```

Testing the case where the child context is cancelled is tougher because `context.WithCancel()` is called inside `connect`, so you don't have access to that `context.CancelFunc`. I see two options here:

1. Spawn a goroutine that calls `connection#closeConnectContext` before dialing. This will block until the child context is fully created and will then cancel it. The Dialer/Handshaker can block waiting for the context to expire as is done in other tests.
2. Alternatively, a simpler implementation would be to just return a `context.Canceled` error from the dialer even though no context has actually expired. This is a bit of a hack but should sufficiently mimic the case where the parent context has not expired, so the pool should be cleared.

matthewdale replied:
I've added a new test `TestServerConnectionCancellation` that tests context cancellation before dialing.

divjotarora reviewed (commented):
Looks great! The new `operationTimeout` parameter for tests is much clearer IMO. I responded to a previous thread about testing `context.Canceled` errors so I'll hold off on approving until that's resolved, but everything else pretty much LGTM.

matthewdale commented:
@kevinAlbs and @divjotarora I've made 2 somewhat significant changes since the last time you reviewed this:
1. Pass the operation-scoped context to the `ProcessHandshakeError` function and check if the actual deadline has passed (if there is one) instead of just checking for `ctx.Err()`. This handles a race condition caused by some net functions that extract the deadline from a context and time out on that deadline, but the goroutine that closes `ctx.Done()` and causes `ctx.Err()` to return an error hasn't run yet.
2. Use individual instances of the new `testPoolMonitor` type to replace the package-global `poolMonitor` in the `mongo/integration` package tests. This fixes unexpected interaction between the pool monitor in different test cases that results in assertion failures and test timeouts (caused by infinitely blocking channel sends).

iwysiu commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,299 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+	eventsWg.Wait()
+	require.NotEmpty(t, events, "expected more than 0 connection pool monitor events")
+
+	// Look for any "ConnectionPoolCleared" events in the events slice so we can assert that
+	// the Server connection pool was or wasn't cleared, depending on the test expectations.
```
```suggestion
	// Look for any "ConnectionPoolCleared" events in the events slice so we can assert that
	// the Server connection pool wasn't cleared.
```

divjotarora commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -54,6 +54,298 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+	}
+}
+
+func TestServerConnectionCancellation(t *testing.T) {
```
nit: For readability, can you write a comment at the beginning of this test that explains what we're testing?

matthewdale replied:
Will add comment

divjotarora reviewed (approved):
LGTM! Thanks for investigating this issue and all of the approaches so thoroughly!

kevinAlbs commented on mongo/integration/primary_stepdown_test.go:
```diff
@@ -23,7 +24,65 @@ const (
 )
 
+// testPoolMonitor exposes an *event.PoolMonitor and collects all events logged to that
+// *event.PoolMonitor. It is safe to use from multiple concurrent goroutines.
```
Oh very nice. Optional: should this be in a separate file, like `connection_pool_helpers_test.go`? Though `isPoolCleared` was used by other files, the name of this file suggests it is only for the tests specified in https://github.com/mongodb/specifications/blob/master/source/connections-survive-step-down/tests/README.rst

matthewdale replied:
Yeah, I wanted to locate it "next to" the existing `poolMonitor` so that there aren't two implementations of the same thing in different places. I think the `testPoolMonitor` could actually be a good candidate for a "test utilities" package because numerous test packages need to record events (including the `Server` tests added in this PR). However, I'm not sure it's worth moving into a separate file until we want to move it into a test utilities package, which probably shouldn't be part of this PR.

kevinAlbs commented on x/mongo/driver/topology/server.go:
```diff
@@ -292,7 +293,29 @@ func (s *Server) ProcessHandshakeError(err, ctxErr error, startingGenerationNumb
+			return true
+		}
+
+		// In some networking functions, the deadline from the context is used to determine timeouts
```
Thank you for the helpful explanation! That is interesting, and sounds like it could be possible buggy behavior in the net package if it inconsistently sets context errors.

kevinAlbs commented on x/mongo/driver/topology/server_test.go:
```diff
@@ -269,6 +270,82 @@ func TestServerConnectionTimeout(t *testing.T) {
+		// Create a connection pool event monitor that sends all events to an events channel
+		// so we can assert on the connection pool events later.
+		WithConnectionPoolMonitor(func(_ *event.PoolMonitor) *event.PoolMonitor {
+			return &event.PoolMonitor{
```
Optional: IIUC this could use the new `testPoolMonitor`.

matthewdale replied:
Yes it could, although it would require moving `testPoolMonitor` into a test util package like `internal/testutil` (or even an external test utilities package if that makes sense). I'll leave a comment to indicate that change should be considered as part of GODRIVER-2068.

kevinAlbs reviewed (approved):
LGTM mod optional comments.
